### ISSUE TITLE
`ty` fails to report type error for incompatible arguments in generic functions, infers union type instead

### ISSUE THREAD

**Issue by WILL-Moxyz** [astral-sh/ty#3431]
### Summary

When a generic function constrains two parameters to be the same type `T`, `ty` does not report a type error if incompatible types are passed. Instead, it infers the return type as the union of the two argument types. This occurs with both PEP 695 syntax and traditional `TypeVar` syntax.

**Minimal reproducible example**

```python
from typing import Protocol, TypeVar

# PEP 695 (Python 3.12+)
class Add[T](Protocol):
    def __add__(self, other: T, /) -> T: ...

def add[T: Add](a: T, b: T, /) -> T:
    return a + b

# Traditional style
T = TypeVar("T", bound="Addable")

class Addable(Protocol):
    def __add__(self, other: T, /) -> T: ...

def _add(a: T, b: T, /) -> T:
    return a + b

def main() -> None:
    r = add(int(1), "123")  # Expected error, but inferred as int | Literal["123"]
    s = add(1, 2)           # Expected int, but inferred as Literal[1, 2]
    t = _add(1, "abc")      # Expected error, but inferred as Literal[1, "abc"]
    print(r, s, t)

if __name__ == "__main__":
    main()
```

**Actual Behavior**

`ty` infers the following types without emitting any error:
- `add(int(1), "123")` → `int | Literal["123"]`  
- `add(1, 2)` → `Literal[1, 2]`  
- `_add(1, "abc")` → `Literal[1, "abc"]`

No type errors or warnings are reported.

**Expected Behavior**

The signature `add[T: Add](a: T, b: T, /) -> T` explicitly requires that `a` and `b` have the same type `T`. Passing `int` and `str` should trigger a type error (e.g., `Argument of type "Literal['123']" cannot be assigned to parameter "b" of type "int"`). The return type should not degenerate into a union of incompatible types.

The same applies to the traditional `TypeVar` version.

**Environment**

- `ty` version: `0.0.34`
- `ruff` version: `0.15.12`
- Python version: `3.12` (as specified by `requires-python = ">=3.12"`)
- Project configuration (`pyproject.toml`):
  ```toml
  [project]
  name = "generic"
  version = "0.1.0"
  requires-python = ">=3.12"
  dependencies = [
      "ruff>=0.15.12",
      "ty>=0.0.34",
  ]
  ```

**Additional Notes**

- The issue appears regardless of whether the generic function is called with explicit type parameters or not.
- Both PEP 695 and traditional syntax are affected, suggesting the constraint solver in `ty` is overly permissive in unifying `TypeVar` values.
- Ideally, `ty` should at least offer a strict mode or flag that makes such violations hard errors.
(Note: I'm a beginner from China; this translation was generated by AI.)

### Version

0.0.34

**Comment by carljm** [comment]
That's not how generics in Python work. `int | str` is a valid solution for `r` and `t`. `T` must solve to some type that is consistent across the call, but that type can be a union type. All Python type checkers agree on this, and it is fundamental to the way Python generic solvers work.

There is one error in your example, in the "legacy" version: `class Addable` needs to inherit `Protocol[T]` so that it is a protocol generic in `T` (like in the PEP 695 version), rather than being a non-generic protocol with a generic method `__add__`. Once that change is made, pyright/mypy also agree that this code is fine. The revealed types differ: some type checkers prefer `int | str`, ty prefers more precise literal types, mypy prefers `Add[Any] / Addable[Any]`, but all of these are valid solutions for `T` in your example.

The corrected code:

```py
from typing import Protocol, TypeVar

# PEP 695 (Python 3.12+)
class Add[T](Protocol):
    def __add__(self, other: T, /) -> T: ...

def add[T: Add](a: T, b: T, /) -> T:
    return a + b

# Traditional style
T = TypeVar("T", bound="Addable")

class Addable(Protocol[T]):
    def __add__(self, other: T, /) -> T: ...

def _add(a: T, b: T, /) -> T:
    return a + b

def main() -> None:
    r = add(int(1), "123")
    reveal_type(r)
    s = add(1, 2)
    reveal_type(s)
    t = _add(1, "abc")
    reveal_type(t)
```

**Comment by WILL-Moxyz** [comment]
Thank you for the detailed explanation! @carljm
